[berkeleydb] build failure

**Host Environment**
- OS: [Microsoft Windows [Version 10.0.19041.423]]
- Compiler: Visual Studio 2019 Version: **16.7.0**  VS Tools Version: **14.27.29110**

**To Reproduce**
Steps to reproduce the behavior:
upgrade Visual Studio 2019 to 16.7.0
`cd [vcpkg-root]`
`git pull` update vcpkg latest as of 20200806 @18:40GMT 14:40EDT
`set VCPKG_DEFAULT_TRIPLET=x64-windows`
`set VCPKG_VISUAL_STUDIO_PATH=C:\Program Files (x86)\Microsoft Visual Studio\2019\Community`
`.\bootstrap-vcpkg.bat -verbose -disableMetrics -win64 -withVSPath "'%VCPKG_VISUAL_STUDIO_PATH%'" `
`.\vcpkg.exe install berkeleydb --editable`

**Failure logs** 
- Cut and past the appropriate build messages from the console output.
`Computing installation plan...`
`The following packages will be built and installed:`
`    berkeleydb[core]:x64-windows`
`Detecting compiler hash for triplet x64-windows...`
`Starting package 1/1: berkeleydb:x64-windows`
`Building package berkeleydb[core]:x64-windows...`
`-- Using cached D:/Git/vcpkg.downloads/db-4.8.30.NC.zip`
`-- Using source at D:/Git/vs2019.prod/buildtrees/berkeleydb/src/db-4-c2cacfc4e9`
`-- Configuring x64-windows`
`-- Building x64-windows-dbg`
`CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:139 (message):`
`    Command failed: "C:/Program Files/CMake/bin/cmake.exe" --build . --config Debug --target install -- -v`
`    Working Directory: D:/Git/vs2019.prod/buildtrees/berkeleydb/x64-windows-dbg`
`    See logs for more information:`
`      D:\Git\vs2019.prod\buildtrees\berkeleydb\install-x64-windows-dbg-out.log`

`Call Stack (most recent call first):`
`  scripts/cmake/vcpkg_build_cmake.cmake:91 (vcpkg_execute_build_process)`
`  scripts/cmake/vcpkg_install_cmake.cmake:24 (vcpkg_build_cmake)`
`  ports/berkeleydb/portfile.cmake:22 (vcpkg_install_cmake)`
`  scripts/ports.cmake:79 (include)`

`Error: Building package berkeleydb:x64-windows failed with: BUILD_FAILED`
`Please ensure you're using the latest portfiles with `.\vcpkg update`, then`
`submit an issue at https://github.com/Microsoft/vcpkg/issues including:`
`  Package: berkeleydb:x64-windows`
`  Vcpkg version: 2020.06.15-nohash`

`Additionally, attach any relevant sections from the log files above.`

- Please attach any additional failure logs mentioned in the console output.
[install-x64-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/5038188/install-x64-windows-dbg-out.log)

**Additional context**
- install-x64-windows-dbg-out.log mentions several issues with file C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\include\atomic lines 1674, 1675, 1685, 1686

**Note 1: port berkeleydb was building fine with VS 2019 version 16.6.5, but now is not building with version 16.7.0**
**Note 2: port berkeleydb still builds fine with VS 2017 version 15.9.25**

--- a/ports/berkeleydb/CONTROL
+++ b/ports/berkeleydb/CONTROL
@@ -1,5 +1,6 @@
 Source: berkeleydb
-Version: 4.8.30-4
+Version: 4.8.30
+Port-Version: 5
 Homepage: https://download.oracle.com/
 Description: BDB - A high-performance embedded database for key/value data.
 Supports: !(uwp|linux|osx)

--- a/ports/berkeleydb/fix-conflict-macro.patch
+++ b/ports/berkeleydb/fix-conflict-macro.patch
@@ -1,0 +1,123 @@
+diff --git a/dbinc/atomic.h b/dbinc/atomic.h
+index 0034dcc..2dd5e03 100644
+--- a/dbinc/atomic.h
++++ b/dbinc/atomic.h
+@@ -70,7 +70,7 @@ typedef struct {
+  * These have no memory barriers; the caller must include them when necessary.
+  */
+ #define	atomic_read(p)		((p)->value)
+-#define	atomic_init(p, val)	((p)->value = (val))
++#define	bdb_atomic_init(p, val)	((p)->value = (val))
+ 
+ #ifdef HAVE_ATOMIC_SUPPORT
+ 
+@@ -206,7 +206,7 @@ static inline int __atomic_compare_exchange(
+ #define	atomic_dec(env, p)	(--(p)->value)
+ #define	atomic_compare_exchange(env, p, oldval, newval)		\
+ 	(DB_ASSERT(env, atomic_read(p) == (oldval)),		\
+-	atomic_init(p, (newval)), 1)
++	bdb_atomic_init(p, (newval)), 1)
+ #else
+ #define atomic_inc(env, p)	__atomic_inc(env, p)
+ #define atomic_dec(env, p)	__atomic_dec(env, p)
+diff --git a/mp/mp_fget.c b/mp/mp_fget.c
+index 5fdee5a..452ef17 100644
+--- a/mp/mp_fget.c
++++ b/mp/mp_fget.c
+@@ -617,7 +617,7 @@ alloc:		/* Allocate a new buffer header and data space. */
+ 
+ 		/* Initialize enough so we can call __memp_bhfree. */
+ 		alloc_bhp->flags = 0;
+-		atomic_init(&alloc_bhp->ref, 1);
++		bdb_atomic_init(&alloc_bhp->ref, 1);
+ #ifdef DIAGNOSTIC
+ 		if ((uintptr_t)alloc_bhp->buf & (sizeof(size_t) - 1)) {
+ 			__db_errx(env,
+@@ -911,7 +911,7 @@ alloc:		/* Allocate a new buffer header and data space. */
+ 			MVCC_MPROTECT(bhp->buf, mfp->stat.st_pagesize,
+ 			    PROT_READ);
+ 
+-		atomic_init(&alloc_bhp->ref, 1);
++		bdb_atomic_init(&alloc_bhp->ref, 1);
+ 		MUTEX_LOCK(env, alloc_bhp->mtx_buf);
+ 		alloc_bhp->priority = bhp->priority;
+ 		alloc_bhp->pgno = bhp->pgno;
+diff --git a/mp/mp_mvcc.c b/mp/mp_mvcc.c
+index 34467d2..b604388 100644
+--- a/mp/mp_mvcc.c
++++ b/mp/mp_mvcc.c
+@@ -276,7 +276,7 @@ __memp_bh_freeze(dbmp, infop, hp, bhp, need_frozenp)
+ #else
+ 	memcpy(frozen_bhp, bhp, SSZA(BH, buf));
+ #endif
+-	atomic_init(&frozen_bhp->ref, 0);
++	bdb_atomic_init(&frozen_bhp->ref, 0);
+ 	if (mutex != MUTEX_INVALID)
+ 		frozen_bhp->mtx_buf = mutex;
+ 	else if ((ret = __mutex_alloc(env, MTX_MPOOL_BH,
+@@ -428,7 +428,7 @@ __memp_bh_thaw(dbmp, infop, hp, frozen_bhp, alloc_bhp)
+ #endif
+ 		alloc_bhp->mtx_buf = mutex;
+ 		MUTEX_LOCK(env, alloc_bhp->mtx_buf);
+-		atomic_init(&alloc_bhp->ref, 1);
++		bdb_atomic_init(&alloc_bhp->ref, 1);
+ 		F_CLR(alloc_bhp, BH_FROZEN);
+ 	}
+ 
+diff --git a/mp/mp_region.c b/mp/mp_region.c
+index e6cece9..e3e3382 100644
+--- a/mp/mp_region.c
++++ b/mp/mp_region.c
+@@ -224,7 +224,7 @@ __memp_init(env, dbmp, reginfo_off, htab_buckets, max_nreg)
+ 			     MTX_MPOOL_FILE_BUCKET, 0, &htab[i].mtx_hash)) != 0)
+ 				return (ret);
+ 			SH_TAILQ_INIT(&htab[i].hash_bucket);
+-			atomic_init(&htab[i].hash_page_dirty, 0);
++			bdb_atomic_init(&htab[i].hash_page_dirty, 0);
+ 		}
+ 
+ 		/*
+@@ -269,7 +269,7 @@ __memp_init(env, dbmp, reginfo_off, htab_buckets, max_nreg)
+ 		hp->mtx_hash = (mtx_base == MUTEX_INVALID) ? MUTEX_INVALID :
+ 		    mtx_base + i;
+ 		SH_TAILQ_INIT(&hp->hash_bucket);
+-		atomic_init(&hp->hash_page_dirty, 0);
++		bdb_atomic_init(&hp->hash_page_dirty, 0);
+ #ifdef HAVE_STATISTICS
+ 		hp->hash_io_wait = 0;
+ 		hp->hash_frozen = hp->hash_thawed = hp->hash_frozen_freed = 0;
+diff --git a/mutex/mut_method.c b/mutex/mut_method.c
+index 2588763..56b6723 100644
+--- a/mutex/mut_method.c
++++ b/mutex/mut_method.c
+@@ -426,7 +426,7 @@ atomic_compare_exchange(env, v, oldval, newval)
+ 	MUTEX_LOCK(env, mtx);
+ 	ret = atomic_read(v) == oldval;
+ 	if (ret)
+-		atomic_init(v, newval);
++		bdb_atomic_init(v, newval);
+ 	MUTEX_UNLOCK(env, mtx);
+ 
+ 	return (ret);
+diff --git a/mutex/mut_tas.c b/mutex/mut_tas.c
+index f3922e0..934a654 100644
+--- a/mutex/mut_tas.c
++++ b/mutex/mut_tas.c
+@@ -46,7 +46,7 @@ __db_tas_mutex_init(env, mutex, flags)
+ 
+ #ifdef HAVE_SHARED_LATCHES
+ 	if (F_ISSET(mutexp, DB_MUTEX_SHARED))
+-		atomic_init(&mutexp->sharecount, 0);
++		bdb_atomic_init(&mutexp->sharecount, 0);
+ 	else
+ #endif
+ 	if (MUTEX_INIT(&mutexp->tas)) {
+@@ -486,7 +486,7 @@ __db_tas_mutex_unlock(env, mutex)
+ 			F_CLR(mutexp, DB_MUTEX_LOCKED);
+ 			/* Flush flag update before zeroing count */
+ 			MEMBAR_EXIT();
+-			atomic_init(&mutexp->sharecount, 0);
++			bdb_atomic_init(&mutexp->sharecount, 0);
+ 		} else {
+ 			DB_ASSERT(env, sharecount > 0);
+ 			MEMBAR_EXIT();

--- a/ports/berkeleydb/portfile.cmake
+++ b/ports/berkeleydb/portfile.cmake
@@ -9,7 +9,7 @@ vcpkg_download_distfile(ARCHIVE
 vcpkg_extract_source_archive_ex(
     OUT_SOURCE_PATH SOURCE_PATH
     ARCHIVE ${ARCHIVE}
-	PATCHES fix-conflict-macro.patch
+    PATCHES fix-conflict-macro.patch
 )
 
 file(COPY ${CMAKE_CURRENT_LIST_DIR}/CMakeLists.txt DESTINATION ${SOURCE_PATH})

--- a/ports/berkeleydb/portfile.cmake
+++ b/ports/berkeleydb/portfile.cmake
@@ -9,6 +9,7 @@ vcpkg_download_distfile(ARCHIVE
 vcpkg_extract_source_archive_ex(
     OUT_SOURCE_PATH SOURCE_PATH
     ARCHIVE ${ARCHIVE}
+	PATCHES fix-conflict-macro.patch
 )
 
 file(COPY ${CMAKE_CURRENT_LIST_DIR}/CMakeLists.txt DESTINATION ${SOURCE_PATH})

